src/main/java/com/myFirstProject/myFirstProject/service/BasketServiceImpl.java

```diff
@@ -34,23 +38,46 @@
     @Autowired
     private ReqConverterService reqConverterService;
 
+    @Autowired
+    private PromoCodeRepository promoCodeRepository;
+
     @Override
     @Transactional
     public Long saveBasket(BasketReq basketReq) {
         logger.info("Get order to update {}", basketReq);
         sumValidation(basketReq);
+        PromoCode promoCode = getPromoCodeFromDB(basketReq);
         Basket order = reqConverterService.convert(basketReq);
+        order.setPromoCode(promoCode);
         LocalDateTime time = LocalDateTime.now();
         order.setExpiredTime(time.plusDays(1L));
-        Basket order1 = basketRepository.save(order);
-        Long id = order1.getId();
+        Basket savedOrder = basketRepository.save(order);
+        Long id = savedOrder.getId();
         logger.info("Basket saved with id {}", id);
-        accountService.payForArticles(basketReq.getSum(), basketReq.getUser().getId());
+        accountService.payForArticles(basketReq.getSum(), basketReq.getUser().getId(), promoCode);
+        promoCode.setValid(false);
         logger.info(String.format("%s was payed for order", basketReq.getSum()));
 
         return id;
     }
 
+    private PromoCode getPromoCodeFromDB(BasketReq basketReq) {
+        PromoCode promoCode = getPromoCode(basketReq);
+        if (promoCode == null) {
```

Comment:
if promoCode == null you received NPE before, see comments for getPromoCode method

Reply:
----